pkg/trace/stats/concentrator_test.go

```diff
@@ -381,42 +431,121 @@ func TestConcentratorSublayersStatsCounts(t *testing.T) {
 		"query|hits|env:none,resource:resource2,service:A2":                                               1,
 		"query|hits|env:none,resource:resource3,service:A2":                                               1,
 		"query|hits|env:none,resource:resource4,service:A3":                                               1,
+		"query|hits|env:none,resource:resource5,service:A3":                                               1,
 		"query|hits|env:none,resource:resource6,service:A3":                                               1,
+		"nested_op|hits|env:none,resource:resource1,service:A1":                                           1,
+		"query|errors|env:none,resource:resource1,service:A1":                                             0,
+		"query|errors|env:none,resource:resource2,service:A2":                                             0,
+		"query|errors|env:none,resource:resource3,service:A2":                                             0,
+		"query|errors|env:none,resource:resource4,service:A3":                                             0,
+		"query|errors|env:none,resource:resource5,service:A3":                                             0,
+		"query|errors|env:none,resource:resource6,service:A3":                                             0,
+		"nested_op|errors|env:none,resource:resource1,service:A1":                                         0,
 	}
+	countsEq(assert, expectedCountValByKey, receivedCounts)
+}
 
-	// verify we got all counts
-	assert.Equal(len(expectedCountValByKey), len(receivedCounts), "GOT %v", receivedCounts)
-	// verify values
-	for key, val := range expectedCountValByKey {
-		count, ok := receivedCounts[key]
-		assert.True(ok, "%s was expected from concentrator", key)
-		assert.Equal(val, int64(count.Value), "Wrong value for count %s", key)
-	}
+func TestConcentrator_AddNow(t *testing.T) {
+	assert := assert.New(t)
+	now := time.Now().UnixNano()
+
+	t.Run("only top-level root span", func(t *testing.T) {
+		statsChan := make(chan []Bucket)
+		trace := pb.Trace{
+			testSpan(1, 0, 50, 5, "query", "A1", "resource1", 0, nil),
+			testSpan(2, 1, 40, 4, "query", "A1", "resource1", 1, nil),
+		}
+		traceutil.ComputeTopLevel(trace)
+		wt := NewWeightedTrace(trace, traceutil.GetRoot(trace))
+		testTrace := &Input{
+			Env:   "none",
+			Trace: wt,
+		}
+		c := NewConcentrator([]string{}, testBucketInterval, statsChan)
+		c.addNow(testTrace, time.Now().UnixNano())
+		expected := map[string]float64{
+			"query|duration|env:none,resource:resource1,service:A1": 50,
+			"query|hits|env:none,resource:resource1,service:A1":     1,
+			"query|errors|env:none,resource:resource1,service:A1":   0,
+		}
+		// skip ahead to first possible flush
+		stats := c.flushNow(now + (int64(c.bufferLen) * testBucketInterval))
+		countsEq(assert, expected, stats[0].Counts)
+	})
+
+	t.Run("top-level root span also marked as measured", func(t *testing.T) {
+		statsChan := make(chan []Bucket)
```

Comment:
These three subtests seem like 90% copy pasta. Let's rewrite them in a way such that there is no duplication. Try something like https://github.com/golang/go/wiki/TableDrivenTests